### ISSUE TITLE
net/http: make Transport retry more aggressively on failures if GetBody is set

### ISSUE THREAD

**Issue by bradfitz** [golang/go#17844]
Make both the http1 and http2 Transports more aggressive on idempotent retries, if the new Go 1.8 Request.GetBody is defined. That means we can back up the Body and replay the whole request.

Go 1.8 uses that for redirects, but not for connection failures.

Do that in Go 1.9.

**Comment by kavirajk** [comment]
can I work on this? @bradfitz

**Comment by bradfitz** [comment]
Sure, if you're familiar with the http code. Note that we already did this for the http2 Transport (during server graceful shutdown in GOAWAY), so be sure to read that code first.

**Comment by kavirajk** [comment]
@bradfitz Thanks. I will spend some time this weekend.

**Comment by kavirajk** [comment]
Sorry for very late reply. 

I started going through the code, to understand the problem correctly. 

I noticed like you mentioned, http2 transport already creating the retry Request via `Request.GetBody` in `http2shouldRetryRequest` function inside `h2_bundle.go`

At the same time, in transport.go, the same operation is done via 
`pconn.shouldRetryRequest()`
 and 
`treq := &transportRequest{Request: req, trace: trace}`

creating new request every time. Instead we have use to Request.GetBody to achieve that.

@bradfitz
